src/fides/api/util/tcf_util.py

```diff
@@ -0,0 +1,637 @@
+                is_purpose_type=is_purpose_section,
+                use_or_feature=attribute,
+            )
+
+            if not top_level_tcf_record:
+                continue
+
+            vendor_id, system_identifier = get_system_identifiers(
+                privacy_declaration_row
+            )
+            # Add top-level entry to the system section if applicable
+            if system_identifier not in system_map:
+                system_map[system_identifier] = TCFVendorRecord(
+                    id=system_identifier,  # Identify system by vendor id if it exists, otherwise use system id.
+                    name=privacy_declaration_row.system_name,
+                    description=privacy_declaration_row.system_description,
+                    has_vendor_id=bool(
+                        vendor_id
+                    ),  # Has_vendor_id will let us separate data into two sections: "tcf_vendors" and "tcf_systems"
+                )
+
+            # Add the purpose or feature to the System purposes or System features lists
+            embedded_purpose_or_feature_record: TCFPurposeOrFeature = (
+                _clone_top_level_record_then_add_legal_bases(
+                    top_level_tcf_record,
+                    privacy_declaration_row.legal_basis_for_processing,
+                )
+            )
+
+            _embed_purpose_or_feature_under_system(
+                embedded_tcf_record=embedded_purpose_or_feature_record,
+                system_section=getattr(
+                    system_map[system_identifier], tcf_component_name
+                ),
+                legal_basis_for_processing=privacy_declaration_row.legal_basis_for_processing,
+            )
+
+            # Do the reverse, and nest the system beneath the purpose or feature
+            _embed_system_under_purpose_or_feature(
+                top_level_tcf_record=top_level_tcf_record,
+                matching_purpose_or_feature_map=matching_purpose_or_feature_map,
+                privacy_declaration_row=privacy_declaration_row,
+            )
+
+    return matching_purpose_or_feature_map, system_map
+
+
+def _extend_legal_bases(
+    purpose_record: TCFPurposeOrFeature,
+    legal_basis_for_processing: Optional[str],
+    replace: bool = False,
+) -> None:
+    """Either appends or replaces the legal_bases, depending on whether replace=True or False
+
+    purpose record updated in place
+    """
+    if not isinstance(purpose_record, TCFPurposeRecord):
+        return
+
+    legal_bases: List[str] = (
+        [legal_basis_for_processing] if legal_basis_for_processing else []
+    )
+
+    if replace:
+        purpose_record.legal_bases = legal_bases
+    else:
+        if legal_basis_for_processing not in purpose_record.legal_bases:
+            purpose_record.legal_bases.extend(legal_bases)
+
+    purpose_record.legal_bases.sort()
+
+
+@lru_cache()
+def get_tcf_contents(
```

Comment:
TODO for me in review -- high-level, this looks great. but there's a lot to dig into here and i haven't had the time to go thru in the level of detail i'd want to.